File: balancin/matlab/sim_hinf.m
clc; clear; close all;               % #3 H-INFINITO (sensibilidad mixta)
% Diseno H-infinito para el subsistema angular inestable del balancin.
% Usa exactamente los ponderadores solicitados y entrega K(z) para firmware.

[A,B,C,D]=planta(); %#ok<ASGLU>
a43=104.6806;
b4=B(4);
Ts=0.01;
HGAIN=0.10;                         % misma escala global usada en hardware

s=tf('s');
G=b4/(s^2-a43);                     % subsistema angulo: theta/u (inestable)

% Ponderadores de sensibilidad mixta
W1=makeweight(200,60,0.05);         % S: desempeno / rechazo error
W2=1e-3;                            % KS: esfuerzo de control
W3=makeweight(0.20,40,2);           % T: robustez / ruido alta frecuencia

% Sintesis H-infinito por sensibilidad mixta
[K,CL,gamma]=mixsyn(G,W1,W2,W3); %#ok<ASGLU>
fprintf('gamma=%.4f\n',gamma);

% Lazo cerrado de referencia y sensibilidad
T=feedback(G*K,1);
S=feedback(1,G*K);
KS=feedback(K,G);

figure('Color','w','Position',[80 80 900 620]);
subplot(2,2,1); step(T); grid on; title('H-inf: T=GK/(1+GK)'); ylabel('\theta/ref');
subplot(2,2,2); bodemag(S,T); grid on; legend('S','T'); title('Sensibilidad y robustez');
subplot(2,2,3); bodemag(KS); grid on; title('Esfuerzo KS');
subplot(2,2,4); pzmap(T); grid on; title('Polos/cero lazo cerrado');
sgtitle(sprintf('H-infinito sensibilidad mixta, gamma=%.3f',gamma));

% Discretizacion para firmware / simulacion
Kd=c2d(K,Ts,'tustin');
[num,den]=tfdata(Kd,'v');
num=real(num); den=real(den);
num=num/den(1); den=den/den(1);

fprintf('\nK(z) Tustin Ts=%.3f s\n',Ts);
fprintf('num = ['); fprintf(' %.9g',num); fprintf(' ];\n');
fprintf('den = ['); fprintf(' %.9g',den); fprintf(' ];\n');
fprintf('HGAIN = %.3f\n',HGAIN);

% Prueba discreta equivalente desde theta0=5 deg sobre el mismo modelo lineal
R2D=180/pi; Tf=4; N=round(Tf/Ts);
sysd=c2d(ss(A,B,eye(4),zeros(4,1)),Ts,'zoh');
Ad=sysd.A; Bd=sysd.B;
x=[0;0;5/R2D;0]; th=zeros(1,N); ulog=zeros(1,N);
nb=numel(num); na=numel(den); ehist=zeros(1,nb); uhist=zeros(1,na);
for k=1:N
    e=-x(3);                         % setpoint theta=0 rad
    ehist=[e ehist(1:end-1)];
    uc=(num*ehist' - den(2:end)*uhist(1:end-1)')/den(1);
    uhist=[uc uhist(1:end-1)];
    u=max(min(HGAIN*uc,255),-255);
    th(k)=x(3)*R2D; ulog(k)=u;
    x=Ad*x+Bd*u;
    if abs(x(3)*R2D)>60
        th=th(1:k); ulog=ulog(1:k); break;
    end
end
t=(0:numel(th)-1)*Ts;
figure('Color','w','Position',[120 120 760 520]);
subplot(2,1,1); plot(t,th,'LineWidth',1.6); grid on; yline(0,'--k');
ylabel('\theta [deg]'); title('H-inf discreto aplicado al modelo completo');
subplot(2,1,2); plot(t,ulog,'LineWidth',1.6); grid on; ylabel('u [PWM]'); xlabel('t [s]');

% Guardar coeficientes para copiar a Arduino si se vuelve a sintetizar K
out=fullfile(fileparts(mfilename('fullpath')),'..','data','coeficientes_hinf_mixsyn.txt');
fid=fopen(out,'w');
fprintf(fid,'// H-INF mixsyn, Ts=%.3f, HGAIN=%.3f, gamma=%.6f\n',Ts,HGAIN,gamma);
fprintf(fid,'const float num[%d]={',nb); fprintf(fid,' %.9gf,',num); fprintf(fid,' };\n');
fprintf(fid,'const float den[%d]={',na); fprintf(fid,' %.9gf,',den); fprintf(fid,' };\n');
fclose(fid);
fprintf('Coeficientes guardados en: %s\n',out);

File: balancin/python/sim_hinf.py
"""
#3  CONTROL H-INFINITO (sensibilidad mixta) para el balancin.
Convencion del curso (PC3/Control_H_infinito): ponderadores W1,W2,W3 + hinfsyn/mixsyn.
Se disena sobre el subsistema de ANGULO (SISO, inestable):
   theta''(t) = 104.68 theta + b4 u  ->  G(s) = b4/(s^2 - 104.68)
El controlador K(s) resulta dinamico -> se discretiza (Tustin) para el Arduino.
"""
import numpy as np, control as ct
import matplotlib; matplotlib.use("Agg"); import matplotlib.pyplot as plt
import os,sys; sys.path.insert(0,os.path.dirname(__file__)); from planta import plant
A,B,C,D=plant(); R2D=180/np.pi

a43=104.6806; b4=float(B[3,0])            # subsistema angulo
s=ct.tf('s'); G=ct.tf([b4],[1,0,-a43])    # planta SISO inestable
print("G(s)=",G)

# --- ponderadores de sensibilidad mixta (tu estilo makeweight/mixsyn) ---
def makeweight(dc,wc,hf):          # peso 1er orden: |W(0)|=dc, |W(inf)|=hf
    s=ct.tf('s'); return (hf*s + wc*dc)/(s + wc)
W1 = makeweight(50, 6, 0.05)       # S: desempeno
W2 = ct.tf([1e-3],[1])             # KS: esfuerzo de control
W3 = makeweight(0.05, 40, 2)       # T: robustez
K,CL,info = ct.mixsyn(G, W1, W2, W3); g_=float(info[0])
print("gamma =", round(g_,3), " (bueno si ~1)")

# --- lazo cerrado, respuesta a condicion inicial (via step de perturbacion) ---
L=ct.series(K,G); T=ct.feedback(L,1)
print("polos lazo cerrado (parte real max):", round(max(np.real(ct.poles(T))),3),
      "-> ESTABLE" if max(np.real(ct.poles(T)))<0 else "-> INESTABLE")

# respuesta: seguir un escalon pequeno de referencia de angulo (0 -> se rechaza dist)
tt=np.linspace(0,3,600)
_,y=ct.step_response(T, tt)
# discretizar el controlador para Arduino (Tustin, dt=10ms)
Kd=ct.tf(ct.c2d(K,0.01,'tustin')); num=np.array(Kd.num[0][0]); den=np.array(Kd.den[0][0])
print("Controlador discreto K(z): num=",np.round(num,5)," den=",np.round(den,5))
plt.figure(figsize=(9,4)); plt.plot(tt,y); plt.grid(alpha=.3)
plt.title(f"H-inf: respuesta lazo cerrado (gamma={g_:.2f})"); plt.xlabel("t[s]"); plt.ylabel("theta/ref")
plt.savefig(os.path.join(os.path.dirname(__file__),"..","data","sim_hinf.png"),dpi=100)
